chain/transaction.go

```diff
@@ -130,39 +132,109 @@ func (t *Transaction) Expiry() int64 { return t.Base.Timestamp }
 
 func (t *Transaction) MaxFee() uint64 { return t.Base.MaxFee }
 
-func (t *Transaction) StateKeys(stateMapping StateManager) (set.Set[string], error) {
-	if t.stateKeys != nil {
-		return t.stateKeys, nil
+func (t *Transaction) ReadKeys(stateMapping StateManager) (set.Set[string], error) {
+	if t.readKeys != nil {
+		return t.readKeys, nil
 	}
 
 	// Verify the formatting of state keys passed by the controller
-	actionKeys := t.Action.StateKeys(t.Auth, t.ID())
-	authKeys := t.Auth.StateKeys()
-	stateKeys := set.NewSet[string](len(actionKeys) + len(authKeys))
-	for _, arr := range [][]string{actionKeys, authKeys} {
+	actionReadKeys := t.Action.ReadKeys(t.Auth, t.ID())
+	authReadKeys := t.Auth.ReadKeys()
+	readKeys := set.NewSet[string](len(actionReadKeys) + len(authReadKeys))
+	for _, arr := range [][]string{actionReadKeys, authReadKeys} {
 		for _, k := range arr {
 			if !keys.Valid(k) {
 				return nil, ErrInvalidKeyValue
 			}
-			stateKeys.Add(k)
+			readKeys.Add(k)
 		}
 	}
 
 	// Add keys used to manage warp operations
 	if t.WarpMessage != nil {
 		p := stateMapping.IncomingWarpKeyPrefix(t.WarpMessage.SourceChainID, t.warpID)
 		k := keys.EncodeChunks(p, MaxIncomingWarpChunks)
-		stateKeys.Add(string(k))
+		readKeys.Add(string(k))
 	}
 	if t.Action.OutputsWarpMessage() {
 		p := stateMapping.OutgoingWarpKeyPrefix(t.id)
 		k := keys.EncodeChunks(p, MaxOutgoingWarpChunks)
-		stateKeys.Add(string(k))
+		readKeys.Add(string(k))
 	}
 
 	// Cache keys if called again
-	t.stateKeys = stateKeys
-	return stateKeys, nil
+	t.readKeys = readKeys
+	return readKeys, nil
+}
+
+func (t *Transaction) WriteKeys(stateMapping StateManager) (set.Set[string], error) {
```

Comment:
`ReadKeys`, `WriteKeys`, and `RWKeys` all have similar functionality, is there a way to prevent duplicate code among these functions?

Reply:
See my comment here: https://github.com/ava-labs/hypersdk/pull/693/files#r1460969811